backends/platform/android/org/scummvm/scummvm/SAFFSTree.java

```diff
@@ -0,0 +1,439 @@
+				}
+
+				SAFFSNode newnode = new SAFFSNode(node, node._path + "/" + displayName, documentId, ourFlags);
+				_cache.put(newnode._path, newnode);
+				results.add(newnode);
+			}
+		} catch (Exception e) {
+			Log.w(ScummVM.LOG_TAG, "Failed query: " + e);
+		} finally {
+			if (c != null) {
+				c.close();
+			}
+		}
+
+		return results.toArray(new SAFFSNode[0]);
+	}
+
+	public SAFFSNode getChild(SAFFSNode node, String name) {
+		final ContentResolver resolver = _context.getContentResolver();
+		final Uri searchUri = DocumentsContract.buildChildDocumentsUriUsingTree(_treeUri, node._documentId);
+
+		Cursor c = null;
+		try {
+			c = resolver.query(searchUri, new String[] { DocumentsContract.Document.COLUMN_DISPLAY_NAME,
+				DocumentsContract.Document.COLUMN_DOCUMENT_ID, DocumentsContract.Document.COLUMN_MIME_TYPE,
+				DocumentsContract.Document.COLUMN_FLAGS }, null, null, null);
+			while (c.moveToNext()) {
+				final String displayName = c.getString(0);
+				if (!name.equals(displayName)) {
+					continue;
+				}
+
+				final String documentId = c.getString(1);
+				final String mimeType = c.getString(2);
+
+				final int flags = c.getInt(3);
+
+				int ourFlags = 0;
+				if (DocumentsContract.Document.MIME_TYPE_DIR.equals(mimeType)) {
+					ourFlags |= SAFFSNode.DIRECTORY;
+				}
+				if ((flags & (DocumentsContract.Document.FLAG_SUPPORTS_WRITE | DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE)) != 0) {
+					ourFlags |= SAFFSNode.WRITABLE;
+				}
+				if ((flags & DocumentsContract.Document.FLAG_VIRTUAL_DOCUMENT) == 0) {
+					ourFlags |= SAFFSNode.READABLE;
+				}
+
+				SAFFSNode newnode = new SAFFSNode(node, node._path + "/" + displayName, documentId, ourFlags);
+				_cache.put(newnode._path, newnode);
+				return newnode;
+			}
+		} catch (Exception e) {
+			Log.w(ScummVM.LOG_TAG, "Failed query: " + e);
+		} finally {
+			if (c != null) {
+				c.close();
+			}
+		}
+
+		_cache.put(node._path + "/" + name, NOT_FOUND_NODE);
+		return null;
+	}
+
+	public SAFFSNode createDirectory(SAFFSNode node, String name) {
+		final ContentResolver resolver = _context.getContentResolver();
+		final Uri parentUri = DocumentsContract.buildDocumentUriUsingTree(_treeUri, node._documentId);
+		Uri newDirUri;
+
+		try {
+			newDirUri = DocumentsContract.createDocument(resolver, parentUri, DocumentsContract.Document.MIME_TYPE_DIR, name);
+		} catch(FileNotFoundException e) {
+			return null;
+		}
+		if (newDirUri == null) {
+			return null;
+		}
+
+		final String documentId = DocumentsContract.getDocumentId(newDirUri);
+
+		final SAFFSNode newnode = new SAFFSNode(node, node._path + "/" + name, documentId, 0);
+		// Update flags
+		final String realName = stat(_root);
+		if (realName == null) {
+			return null;
+		}
+		// Unlikely but...
+		if (!realName.equals(name)) {
+			newnode._path = node._path + "/" + realName;
+		}
+
+		_cache.put(newnode._path, newnode);
+
+		return newnode;
+	}
+
+	public SAFFSNode createFile(SAFFSNode node, String name) {
```

Comment:
Since this code is 99% the same as `createDirectory()` above, perhaps it would make sense to reuse it in an internal funcion?

Reply:
Also done.